### PR TITLE
Synthetics Client update: New fields, new capabilities

### PR DESCRIPTION
**Pull request by greatestusername**
Prepares for new major release of provider (`2.0.0`)

**Major Changes**
- Changes in `syntheticsclient` [version `2.0.2`/`2.0.3`](https://github.com/splunk/syntheticsclient/releases/tag/v2.0.2)
- Correctly utilize pointers to pass nulls for some fields (notably `authentication` and `user_agent`)
- Add new settings for HTTP Checks (authentication, user agent, verify certificates)
- Updates to resource read (and import)
- Add new settings for `validations` and `setup`
- Add option for `wait` type and `duration` on Browser steps
- Add option for `collect_interactive_metrics` on Browser checks
- Updates to docs and examples

### CODE REVIEW

File: synthetics/resource_browser_check_v2.go [hunk 1]
```diff
@@ -80,16 +80,21 @@ func resourceBrowserCheckV2() *schema.Resource {
...
 								Schema: map[string]*schema.Schema{
 									"user_agent": {
 										Type:     schema.TypeString,
 										Optional: true,
 									},
 									"verify_certificates": {
+										Type:     schema.TypeBool,
+										Required: true,
+									},
+									"collect_interactive_metrics": {
```
Comment (greatestusername) [inline synthetics/resource_browser_check_v2.go:94]: This is the main reason for a Major release. 
`advanced_settings` is now required. We want opinionated declaration of `verify_certificates`

Also adds ability to manage `collect_interactive_metrics` for a check

File: synthetics/resource_browser_check_v2.go [hunk 2]
```diff
@@ -231,9 +236,14 @@ func resourceBrowserCheckV2() *schema.Resource {
 													Type:     schema.TypeString,
 													Optional: true,
 												},
+												"duration": {
+													Type:     schema.TypeInt,
+													Optional: true,
+												},
```
Comment (greatestusername) [inline synthetics/resource_browser_check_v2.go:242]: `duration` is now an optional but possible setting for `wait` type steps

File: synthetics/resource_browser_check_v2.go [hunk 3]
```diff
@@ -295,6 +305,9 @@ func resourceBrowserCheckV2Read(ctx context.Context, d *schema.ResourceData, met
 		return diag.FromErr(err)
 	}
 	log.Println("[DEBUG] GET BROWSER BODY: ", o)
+	if err := d.Set("test", flattenBrowserV2Read(o)); err != nil {
```
Comment (greatestusername) [inline synthetics/resource_browser_check_v2.go:308]: @mcmluk-monster Implementing this read as you had in your previous PR but with underlying fields updated.

File: synthetics/resource_http_check_v2.go [hunk 4]
```diff
@@ -83,6 +95,14 @@ func resourceHttpCheckV2() *schema.Resource {
...
 							},
 						},
+						"user_agent": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"verify_certificates": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
```
Comment (greatestusername) [inline synthetics/resource_http_check_v2.go:105]: `user_agent` and `verify_certificates` are now available as options for HTTP checks

File: synthetics/resource_port_check_v2.go [hunk 5]
```diff
@@ -41,6 +41,18 @@ func resourcePortCheckV2() *schema.Resource {
...
+							Computed: true,
+						},
+						"created_at": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"updated_at": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
```
Comment (greatestusername) [inline synthetics/resource_port_check_v2.go:55]: These fields will always be returned by the API and thus computed

File: synthetics/structures.go [hunk 6]
```diff
@@ -44,6 +44,37 @@ func flattenStringIdData(test interface{}) string {
...
+
+	apiV2["active"] = checkApiV2.Test.Active
+
+	if checkApiV2.Test.Frequency != 0 {
+		apiV2["frequency"] = checkApiV2.Test.Frequency
+	}
+
+	if checkApiV2.Test.Name != "" {
+		apiV2["name"] = checkApiV2.Test.Name
+	}
```
Comment (greatestusername) [inline synthetics/structures.go:58]: Implemented read function and updated from a previous PR

File: synthetics/structures.go [hunk 7]
```diff
@@ -352,13 +417,17 @@ func flattenHttpV2Read(checkHttpV2 *sc2.HttpCheckV2Response) []interface{} {
 		httpV2["body"] = checkHttpV2.Test.Body
 	}
 
+	httpV2["user_agent"] = checkHttpV2.Test.UserAgent
+
+	httpV2["verify_certificates"] = checkHttpV2.Test.Verifycertificates
+
```
Comment (greatestusername) [inline synthetics/structures.go:422]: http checks can now have user_agent and verify_certificates

File: synthetics/structures.go [hunk 8]
```diff
@@ -410,6 +479,12 @@ func flattenHttpV2Data(checkHttpV2 *sc2.HttpCheckV2Response) []interface{} {
 		httpV2["body"] = checkHttpV2.Test.Body
 	}
 
+	if *checkHttpV2.Test.UserAgent != "" {
+		httpV2["user_agent"] = checkHttpV2.Test.UserAgent
+	}
+
+	httpV2["verify_certificates"] = checkHttpV2.Test.Verifycertificates
+
```
Comment (greatestusername) [inline synthetics/structures.go:486]: http checks can now have user_agent and verify_certificates

File: synthetics/structures.go [hunk 9]
```diff
@@ -677,8 +748,12 @@ func flattenStepsData(checkSteps *[]sc2.StepsV2) []interface{} {
 				cl["selector_type"] = checkStep.SelectorType
 			}
 
-			options := flattenOptionsData(&checkStep.Options)
-			cl["options"] = options
+			//options := flattenOptionsData(&checkStep.Options)
```
Comment (greatestusername) [inline synthetics/structures.go]: `options` field is deprecated by the API

File: synthetics/structures.go [hunk 10]
```diff
@@ -677,8 +747,9 @@ func flattenStepsData(checkSteps *[]sc2.StepsV2) []interface{} {
 				cl["selector_type"] = checkStep.SelectorType
 			}
 
-			options := flattenOptionsData(&checkStep.Options)
```
Comment (greatestusername) [inline synthetics/structures.go:680]: `options` no longer utilized by api

File: synthetics/structures.go [hunk 11]
```diff
@@ -716,6 +787,14 @@ func flattenSetupData(checkSetup *[]sc2.Setup) []interface{} {
...
 			}
 
+			if checkSetup.Code != "" {
+				cl["code"] = checkSetup.Code
+			}
+
+			if checkSetup.Value != "" {
+				cl["value"] = checkSetup.Value
+			}
+
```
Comment (greatestusername) [inline synthetics/structures.go:797]: `code` and `value` now fields for `setup`

File: synthetics/structures.go [hunk 12]
```diff
@@ -840,6 +919,26 @@ func flattenValidationsData(checkValidations *[]sc2.Validations) []interface{} {
...
+				cl["variable"] = checkValidations.Variable
+			}
+
+			if checkValidations.Value != "" {
+				cl["value"] = checkValidations.Value
+			}
+
+			if checkValidations.Code != "" {
+				cl["code"] = checkValidations.Code
+			}
```
Comment (greatestusername) [inline synthetics/structures.go:940]: New options for validation since last update

File: synthetics/structures.go [hunk 13]
```diff
@@ -909,27 +1008,33 @@ func flattenDeviceData(checkDevice *sc2.Device) []interface{} {
 	return []interface{}{device}
 }
 
-func flattenAdvancedSettingsData(checkDevice *sc2.Advancedsettings) []interface{} {
+func flattenAdvancedSettingsData(advSettings *sc2.Advancedsettings) []interface{} {
```
Comment (greatestusername) [inline synthetics/structures.go:1011]: `checkDevice` was an old copy/pasta issue

File: synthetics/structures.go [hunk 14]
```diff
@@ -909,27 +1008,33 @@ func flattenDeviceData(checkDevice *sc2.Device) []interface{} {
...
-	Authentication := flattenAuthenticationData(&checkDevice.Authentication)
-	advancedSettings["authentication"] = Authentication
+	if advSettings.UserAgent != nil {
+		advancedSettings["user_agent"] = advSettings.UserAgent
+	}
+
+	if advSettings.Authentication != nil {
+		Authentication := flattenAuthenticationData(advSettings.Authentication)
+		advancedSettings["authentication"] = Authentication
+	}
```
Comment (greatestusername) [inline synthetics/structures.go:1029]: `nil` checks due to pointer used in v2 of client to send `null`

File: synthetics/structures.go [hunk 15]
```diff
@@ -1207,8 +1319,10 @@ func buildAdvancedSettingsData(advancedSettings *schema.Set) sc2.Advancedsetting
 	if len(as_list) > 0 {
 		as_map := as_list[0].(map[string]interface{})
 
+		userAgentString := as_map["user_agent"].(string)
+		advancedSettingsData.UserAgent = &userAgentString
```
Comment (greatestusername) [inline synthetics/structures.go:1323]: Best way I could find to manage string to string pointer

File: synthetics/structures.go [hunk 16]
```diff
@@ -1269,16 +1383,18 @@ func buildHostOverridesData(hostOverrides *schema.Set) []sc2.HostOverrides {
...
-	return authenticationData
+	if len(authentication_list) > 0 {
+		authentication_map := authentication_list[0].(map[string]interface{})
+		authenticationData := &sc2.Authentication{
+			Username: authentication_map["username"].(string),
+			Password: authentication_map["password"].(string),
+		}
+		return authenticationData
+	}
+	return nil
```
Comment (greatestusername) [inline synthetics/structures.go:1397]: Handling due to needing possibility to return `nil` and send `null` with client

**Review by greatestusername — COMMENTED** [review]
Added a bunch of review notes.

**Any file with changes other than version bump have comments** use comments to ignore other files while reviewing

File: synthetics/structures.go [hunk 17]
```diff
@@ -1017,6 +1120,9 @@ func buildHttpV2Data(d *schema.ResourceData) sc2.HttpCheckV2Input {
 			httpv2.Test.Active = http["active"].(bool)
 			httpv2.Test.RequestMethod = http["request_method"].(string)
 			httpv2.Test.Body = http["body"].(string)
+			httpv2.Test.Verifycertificates = http["verify_certificates"].(bool)
+			userAgentString := http["user_agent"].(string)
+			httpv2.Test.UserAgent = &userAgentString
```
Comment (greatestusername) [inline synthetics/structures.go:1125]: must read `string` from list but return a pointer

**Review by greatestusername-splunk — APPROVED** [review]

File: synthetics/resource_browser_check_v2.go (hunk 3 above)
Comment (mcmluk-monster) [reply synthetics/resource_browser_check_v2.go:308]: Looks good to me! I had a quick check and ran a bunch of tests and works like a charm! Thanks got getting the changes in!

**Pull request merged**